pele_platform/Utilities/Helpers/map_atoms.py

```diff
@@ -0,0 +1,138 @@
+from typing import Any, List
+
+from pele_platform.Utilities.Helpers import helpers
+from pele_platform.Utilities.Parameters import pele_env
+from pele_platform.constants import constants
+
+
+class AtomMapper:
+    """
+    If atom or residue numbers change during preprocessing, the AtomMapper will map them to the new PDB file based on
+    atomic coordinates. All fields specified in constants.atom_string_flags are checked automatically.
+    """
+
+    def __init__(
+        self,
+        args: pele_env.EnviroBuilder,
+        env: pele_env.EnviroBuilder,
+        original_system: str,
+    ) -> None:
+        self.args = args
+        self.logger = env.logger
+        self.original_system = original_system
+        self.preprocessed_pdb = env.system
+        self.atom_string_flags = constants.atom_string_flags
+        self.all_args = [
+            arg
+            for arg in self.atom_string_flags
+            if getattr(self.args, arg, None) is not None
+        ]
+
+    def run(self) -> pele_env.EnviroBuilder:
+        for arg in self.all_args:
+            arg_value = getattr(self.args, arg)
+            arg_value = atom_number_to_atom_string(self.original_system, arg_value)
+            new_atom_string = self.check_atom_string(arg_value)
+            setattr(self.args, arg, new_atom_string)
+        return self.args
+
+    def check_atom_string(self, args: List[str]) -> List[str]:
+        """
+        Checks if the atom string needs mapping by attempting to extract its coordinates.
+        """
+        output = []
+        args = args if isinstance(args, list) else [args]
+        for arg in args:
+            try:
+                helpers.get_coords_from_residue(self.original_system, arg)
+                helpers.get_coords_from_residue(self.preprocessed_pdb, arg)
+                output.append(arg)
+            except Exception as e:
+                self.logger.info("{} - mapping it now!".format(e))
+                _, after = self.map_atom_string(
+                    arg, self.original_system, self.preprocessed_pdb, self.logger
+                )
+                output.append(after)
+        return output
+
+    @staticmethod
+    def map_atom_string(
+        atom_string: str,
+        original_input: str,
+        preprocessed_file: str,
+        logger=Any,
+    ) -> (str, str):
+        """
+        Maps old atom string to a new atom string by comparing coordinates of the original and preprocessed PBD files.
+        """
+
+        # read in the original and preprocessed PDB lines
+        with open(original_input, "r") as initial:
+            initial_lines = initial.readlines()
+
+        with open(preprocessed_file, "r") as prep:
+            preprocessed_lines = prep.readlines()
+
+        # retrieve atom info from the original PDB
+        chain, resnum, atom_name = atom_string.split(":")
+
+        # extract coordinates from the original PDB
+        for i in initial_lines:
```

Comment:
`i` could be misleading. It is usually used to define an integer. If `i` represents a line, I would call it `line`.